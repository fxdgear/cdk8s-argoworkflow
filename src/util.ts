import * as k8s from './imports/k8s';

export interface InputParams {
  readonly repo_url: string;
  readonly repo_ref: string;
  readonly repo_name?: string;
  readonly pr_number?: string;
  readonly repo_commit_id?: string;
  readonly root_ecr?: string,
}

export interface sshKey {
  readonly name: string;
  readonly key: string;
}

export interface Git {
  readonly repo: string;
  readonly revision?:string;
  readonly sshPrivateKeySecret?: sshKey;
  readonly insecureIgnoreHostKey?: boolean;
}

export interface initContainer {
  readonly name: string;
  readonly image?: string;
  readonly env?: k8s.EnvVar;
  readonly command?: string[];
  readonly args?: string[];
}

export interface Artifact {
  name: string,
  path: string,
  git?: Git
}

export function artifactGenerator(
  path: string, 
  type: string,
  secret?: sshKey
) {
  switch(type) {
    case 'git':
      const artifact: Artifact = {
        name: 'repo-source',
        path: path,
        git: {
          repo: "{{inputs.parameters.repo_url}}",
          revision: "{{inputs.parameters.repo_ref}}",
          sshPrivateKeySecret: {
            name: secret.name,
            key: secret.key
          },
          insecureIgnoreHostKey: false
        }
      }
      return artifact;
  }
}

export function inputGenerator(
  inputParams: InputParams,
  artifactPath?: string,
  secret?: sshKey
) {
  const input: Input = {
    parameters: inputParams,
    artifacts: artifactGenerator(
      artifactPath,
      'git',
      secret
    )
  };
  return input;
}