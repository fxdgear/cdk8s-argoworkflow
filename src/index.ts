import * as cdk8s from 'cdk8s';
import { ApiObject, GroupVersionKind } from 'cdk8s';
import { Construct } from 'constructs';
import * as k8s from './imports/k8s';
import * as swagger from './output';
import { InputParams, initContainer, sshKey } from './util';

export interface WorkflowProps {
  readonly name?: string;
  readonly namespace?: string;
  readonly image?: string;
  readonly volumes?: k8s.Volume;
  readonly templateName?: string;
  readonly inputParams?: swagger.Inputs;
  readonly containerVals?: k8s.Container;
  readonly securityContext?: k8s.SecurityContext;
  readonly initContainers?: k8s.Container;
  readonly gitSshPrivateKeySecret?: sshKey;
  readonly artifactPath?: string;
}

export class ArgoWorkflow extends ApiObject {
  public static readonly GVK: GroupVersionKind = {
    apiVersion: 'argoproj.io/v1alpha1',
    kind: 'ClusterWorkflowTemplate'
  }
  
  public static manifest(props: WorkflowProps): any {
    return {
      ...ArgoWorkflow.GVK,
      ...props
    }
  }

  constructor(scope: Construct, name: string, props: WorkflowProps) {
    super(scope, name, ArgoWorkflow.manifest(props));
  }
}