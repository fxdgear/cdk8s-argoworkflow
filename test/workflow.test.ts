
import { Chart, Testing } from 'cdk8s';
import * as temp from '../src/index';

test('Workflow', () => {
  const app = Testing.app();
  const chart = new Chart(app, 'test');

  new temp.ArgoWorkflowWorkflowTemplate(chart, 'workflow', {
    metadata: {
      name: 'test',
    },
    spec: {
      templates: [{
        name: 'test',
        container: {
          name: 'test',
          image: 'test',
          command: ['test'],
          args: ['test'],
        },
      }],
    },
  });

  expect(Testing.synth(chart)).toMatchSnapshot();
});
