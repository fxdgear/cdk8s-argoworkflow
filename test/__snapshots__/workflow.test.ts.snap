// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Workflow 1`] = `
Array [
  Object {
    "apiVersion": "argoproj.io/v1alpha1",
    "kind": "WorkflowTemplate",
    "metadata": Object {
      "name": "test",
    },
    "spec": Object {
      "templates": Array [
        Object {
          "container": Object {
            "args": Array [
              "test",
            ],
            "command": Array [
              "test",
            ],
            "image": "test",
            "name": "test",
          },
          "name": "test",
        },
      ],
    },
  },
]
`;
